// @volar/client/out/features/splitEditors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.userPick = exports.activate = exports.htmlLs = void 0;
const vscode = require("vscode");
const reactivity_1 = require("@vue/reactivity");
const shared = require("@volar/shared");
const html = require("vscode-html-languageservice");
const vscode_languageserver_textdocument_1 = require("vscode-languageserver-textdocument");
exports.htmlLs = html.getLanguageService();
function activate(context) {
    const getDocDescriptor = useDocDescriptor();
    context.subscriptions.push(vscode.commands.registerCommand('volar.action.splitEditors', onSplit));
    async function onSplit() {
        var _a;
        const editor = vscode.window.activeTextEditor;
        if (!editor)
            return;
        const doc = editor.document;
        const descriptor = getDocDescriptor(doc.getText());
        const blocksSet = [];
        if (descriptor.scriptSetup || descriptor.script) {
            blocksSet.push([descriptor.scriptSetup, descriptor.script].filter(shared.notEmpty));
        }
        if (descriptor.template) {
            blocksSet.push([descriptor.template]);
        }
        if (descriptor.styles.length) {
            blocksSet.push(descriptor.styles);
        }
        if (descriptor.customBlocks.length) {
            blocksSet.push(descriptor.customBlocks);
        }
        const startViewColumn = (_a = vscode.window.activeTextEditor) === null || _a === void 0 ? void 0 : _a.viewColumn;
        if (startViewColumn === undefined)
            return;
        await splitCurrentEditors(blocksSet.length);
        for (let i = 0; i < blocksSet.length; i++) {
            const blocks = blocksSet[i];
            const firstBlock = blocks.sort((a, b) => a.startTagEnd - b.startTagEnd)[0];
            const editor = await vscode.window.showTextDocument(doc, startViewColumn + i);
            editor.selections = blocks.map(block => new vscode.Selection(doc.positionAt(block.startTagEnd), doc.positionAt(block.startTagEnd)));
            await vscode.commands.executeCommand('editor.foldLevel1');
            editor.revealRange(new vscode.Range(doc.positionAt(firstBlock.startTagEnd), new vscode.Position(editor.document.lineCount, 0)), vscode.TextEditorRevealType.AtTop);
        }
    }
}
exports.activate = activate;
async function splitCurrentEditors(num) {
    const actions = [];
    await vscode.commands.executeCommand('editor.unfoldAll');
    for (let i = 1; i < num; i++) {
        if (i % 2 === 1) {
            actions.push(vscode.commands.executeCommand('workbench.action.splitEditorRight'));
        }
        if (i % 2 === 0) {
            actions.push(vscode.commands.executeCommand('workbench.action.splitEditorDown'));
        }
    }
    await Promise.all(actions);
}
function useDocDescriptor() {
    const splitDocText = (0, reactivity_1.ref)('');
    const splitDocDescriptor = (0, reactivity_1.computed)(() => shared.parseSfc(splitDocText.value, exports.htmlLs.parseHTMLDocument(vscode_languageserver_textdocument_1.TextDocument.create('', '', 0, splitDocText.value))));
    return getDescriptor;
    function getDescriptor(text) {
        splitDocText.value = text;
        return splitDocDescriptor.value;
    }
}
function userPick(options, placeholder) {
    return new Promise(resolve => {
        const quickPick = vscode.window.createQuickPick();
        quickPick.items = Object.values(options);
        quickPick.placeholder = placeholder;
        quickPick.onDidChangeSelection(selection => {
            if (selection[0]) {
                for (let key in options) {
                    const option = options[key];
                    if (selection[0] === option) {
                        resolve(key);
                        quickPick.hide();
                        break;
                    }
                }
            }
        });
        quickPick.onDidHide(() => {
            quickPick.dispose();
            resolve(undefined);
        });
        quickPick.show();
    });
}
exports.userPick = userPick;
