// @volar/server/out/projects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createProjects = void 0;
const shared = require("@volar/shared");
const upath = require("upath");
const vscode = require("vscode-languageserver");
const project_1 = require("./project");
function createProjects(options, ts, tsLocalized, connection, documents, rootPaths, inferredCompilerOptions, lsConfigs) {
    let semanticTokensReq = 0;
    let documentUpdatedReq = 0;
    const updatedUris = new Set();
    const tsConfigNames = ['tsconfig.json', 'jsconfig.json'];
    const projects = new Map();
    const inferredProjects = new Map();
    const progressMap = new Map();
    const inferredProgressMap = new Map();
    initProjects();
    documents.onDidChangeContent(async (change) => {
        for (const [_, service] of [...projects, ...inferredProjects]) {
            await service.onDocumentUpdated(change.document);
        }
        updateDiagnostics(change.document.uri);
    });
    documents.onDidClose(change => connection.sendDiagnostics({ uri: change.document.uri, diagnostics: [] }));
    connection.onDidChangeWatchedFiles(async (handler) => {
        var _a, _b;
        const tsConfigChanges = [];
        const scriptChanges = [];
        for (const change of handler.changes) {
            const fileName = shared.uriToFsPath(change.uri);
            if (tsConfigNames.includes(upath.basename(fileName))) {
                tsConfigChanges.push(change);
            }
            else {
                scriptChanges.push(change);
            }
        }
        if (tsConfigChanges.length) {
            clearDiagnostics();
            const { projectProgress } = await getProjectProgress(tsConfigChanges.map(change => shared.uriToFsPath(change.uri)), []);
            for (const tsConfigChange of tsConfigChanges) {
                const tsConfig = shared.uriToFsPath(tsConfigChange.uri);
                if (tsConfigChange.type === vscode.FileChangeType.Deleted) {
                    if (projects.has(tsConfig)) {
                        (_a = projects.get(tsConfig)) === null || _a === void 0 ? void 0 : _a.dispose();
                        projects.delete(tsConfig);
                    }
                }
                else {
                    setProject(tsConfig, projectProgress[tsConfig]);
                }
            }
        }
        if (scriptChanges.length) {
            // fix sometime vscode file watcher missing tsconfigs delete changes
            for (const tsConfig of projects.keys()) {
                if (!ts.sys.fileExists(tsConfig)) {
                    (_b = projects.get(tsConfig)) === null || _b === void 0 ? void 0 : _b.dispose();
                    projects.delete(tsConfig);
                }
            }
            for (const [_, project] of [...projects, ...inferredProjects]) {
                await project.onWorkspaceFilesChanged(scriptChanges);
            }
        }
        onDriveFileUpdated(undefined);
    });
    return {
        projects,
        inferredProjects,
        get,
    };
    async function initProjects() {
        const tsConfigs = searchTsConfigs();
        const { projectProgress } = await getProjectProgress(tsConfigs, []);
        for (const tsConfig of tsConfigs) {
            setProject(tsConfig, projectProgress[tsConfig]);
        }
        updateInferredProjects();
        updateDiagnostics(undefined);
    }
    function searchTsConfigs() {
        const tsConfigSet = new Set(rootPaths
            .map(rootPath => ts.sys.readDirectory(rootPath, tsConfigNames, undefined, ['**/*']))
            .flat()
            .filter(tsConfig => tsConfigNames.includes(upath.basename(tsConfig))));
        const tsConfigs = [...tsConfigSet];
        return tsConfigs;
    }
    async function getProjectProgress(tsconfigs, inferredTsconfigs) {
        for (const tsconfig of tsconfigs) {
            if (!progressMap.has(tsconfig)) {
                progressMap.set(tsconfig, connection.window.createWorkDoneProgress());
            }
        }
        for (const tsconfig of inferredTsconfigs) {
            if (!inferredProgressMap.has(tsconfig)) {
                inferredProgressMap.set(tsconfig, connection.window.createWorkDoneProgress());
            }
        }
        const projectProgress = {};
        const inferredProjectProgress = {};
        for (const tsconfig of tsconfigs) {
            const progress = progressMap.get(tsconfig);
            if (progress) {
                projectProgress[tsconfig] = await progress;
            }
        }
        for (const tsconfig of inferredTsconfigs) {
            const progress = inferredProgressMap.get(tsconfig);
            if (progress) {
                inferredProjectProgress[tsconfig] = await progress;
            }
        }
        return {
            projectProgress,
            inferredProjectProgress,
        };
    }
    async function updateDiagnostics(docUri) {
        var _a;
        if (!((_a = options.languageFeatures) === null || _a === void 0 ? void 0 : _a.diagnostics))
            return;
        if (docUri) {
            updatedUris.add(docUri);
        }
        const req = ++documentUpdatedReq;
        await shared.sleep(100);
        if (req !== documentUpdatedReq)
            return;
        const changeDocs = [...updatedUris].map(uri => documents.get(uri)).filter(shared.notEmpty);
        const otherDocs = documents.all().filter(doc => !updatedUris.has(doc.uri));
        for (const changeDoc of changeDocs) {
            if (req !== documentUpdatedReq)
                return;
            let _isCancel = false;
            const isDocCancel = getCancelChecker(changeDoc.uri, changeDoc.version);
            const isCancel = async () => {
                const result = req !== documentUpdatedReq || await isDocCancel();
                _isCancel = result;
                return result;
            };
            await sendDocumentDiagnostics(changeDoc.uri, isCancel);
            if (!_isCancel) {
                updatedUris.delete(changeDoc.uri);
            }
        }
        for (const doc of otherDocs) {
            if (req !== documentUpdatedReq)
                return;
            const changeDoc = docUri ? documents.get(docUri) : undefined;
            const isDocCancel = changeDoc ? getCancelChecker(changeDoc.uri, changeDoc.version) : async () => {
                await shared.sleep(0);
                return false;
            };
            const isCancel = async () => req !== documentUpdatedReq || await isDocCancel();
            ;
            await sendDocumentDiagnostics(doc.uri, isCancel);
        }
    }
    function getCancelChecker(uri, version) {
        let _isCancel = false;
        let lastResultAt = Date.now();
        return async () => {
            var _a;
            if (_isCancel) {
                return true;
            }
            if (typeof ((_a = options.languageFeatures) === null || _a === void 0 ? void 0 : _a.diagnostics) === 'object'
                && options.languageFeatures.diagnostics.getDocumentVersionRequest
                && Date.now() - lastResultAt >= 1 // 1ms
            ) {
                const clientDocVersion = await connection.sendRequest(shared.GetDocumentVersionRequest.type, { uri });
                if (clientDocVersion !== null && clientDocVersion !== undefined && version !== clientDocVersion) {
                    _isCancel = true;
                }
                lastResultAt = Date.now();
            }
            return _isCancel;
        };
    }
    async function sendDocumentDiagnostics(uri, isCancel) {
        const match = get(uri);
        if (!match)
            return;
        let send = false; // is vue document
        await match.service.doValidation(uri, async (result) => {
            send = true;
            connection.sendDiagnostics({ uri: uri, diagnostics: result });
        }, isCancel);
    }
    function setProject(tsConfig, progress) {
        var _a;
        if (projects.has(tsConfig)) {
            (_a = projects.get(tsConfig)) === null || _a === void 0 ? void 0 : _a.dispose();
            projects.delete(tsConfig);
        }
        projects.set(tsConfig, (0, project_1.createProject)(ts, options, upath.dirname(tsConfig), tsConfig, tsLocalized, documents, progress, connection, lsConfigs));
    }
    async function updateInferredProjects() {
        const inferredTsConfigs = rootPaths.map(rootPath => upath.join(rootPath, 'tsconfig.json'));
        const inferredTsConfigsToCreate = inferredTsConfigs.filter(tsConfig => !inferredProjects.has(tsConfig));
        const { inferredProjectProgress } = await getProjectProgress([], inferredTsConfigsToCreate);
        for (const inferredTsConfig of inferredTsConfigsToCreate) {
            inferredProjects.set(inferredTsConfig, (0, project_1.createProject)(ts, options, upath.dirname(inferredTsConfig), inferredCompilerOptions, tsLocalized, documents, inferredProjectProgress[inferredTsConfig], connection, lsConfigs));
        }
    }
    async function onDriveFileUpdated(driveFileName) {
        var _a;
        const req = ++semanticTokensReq;
        await updateDiagnostics(driveFileName ? shared.fsPathToUri(driveFileName) : undefined);
        await shared.sleep(100);
        if (req === semanticTokensReq) {
            if ((_a = options.languageFeatures) === null || _a === void 0 ? void 0 : _a.semanticTokens) {
                connection.languages.semanticTokens.refresh();
            }
        }
    }
    function clearDiagnostics() {
        for (const doc of documents.all()) {
            connection.sendDiagnostics({ uri: doc.uri, diagnostics: [] });
        }
    }
    function get(uri) {
        const tsConfigs = getMatchTsConfigs(uri, projects);
        if (tsConfigs.length) {
            const service = projects.get(tsConfigs[0]);
            if (service) {
                return {
                    tsConfig: tsConfigs[0],
                    service: service.getLanguageService(),
                };
            }
        }
        const inferredTsConfigs = getMatchTsConfigs(uri, inferredProjects);
        if (inferredTsConfigs.length) {
            const service = inferredProjects.get(inferredTsConfigs[0]);
            if (service) {
                return {
                    tsConfig: inferredTsConfigs[0],
                    service: service.getLanguageService(),
                };
            }
        }
    }
    function getMatchTsConfigs(uri, services) {
        var _a;
        const fileName = shared.uriToFsPath(uri);
        let firstMatchTsConfigs = [];
        let secondMatchTsConfigs = [];
        for (const kvp of services) {
            const tsConfig = upath.resolve(kvp[0]);
            const parsedCommandLine = kvp[1].getParsedCommandLine();
            const fileNames = new Set(parsedCommandLine.fileNames);
            if (fileNames.has(fileName) || ((_a = kvp[1].getLanguageServiceDontCreate()) === null || _a === void 0 ? void 0 : _a.__internal__.context.scriptTsLs.__internal__.getValidTextDocument(uri))) {
                const tsConfigDir = upath.dirname(tsConfig);
                if (!upath.relative(tsConfigDir, fileName).startsWith('..')) { // is file under tsconfig.json folder
                    firstMatchTsConfigs.push(tsConfig);
                }
                else {
                    secondMatchTsConfigs.push(tsConfig);
                }
            }
        }
        firstMatchTsConfigs = firstMatchTsConfigs.sort(sortPaths);
        secondMatchTsConfigs = secondMatchTsConfigs.sort(sortPaths);
        return [
            ...firstMatchTsConfigs,
            ...secondMatchTsConfigs,
        ];
        function sortPaths(a, b) {
            const aLength = a.split('/').length;
            const bLength = b.split('/').length;
            if (aLength === bLength) {
                const aWeight = upath.basename(a) === 'tsconfig.json' ? 1 : 0;
                const bWeight = upath.basename(b) === 'tsconfig.json' ? 1 : 0;
                return bWeight - aWeight;
            }
            return bLength - aLength;
        }
    }
}
exports.createProjects = createProjects;
