// @volar/client/out/features/autoInsertion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.activate = void 0;
const vscode = require("vscode");
const shared = require("@volar/shared");
async function activate(context, htmlClient, tsClient) {
    await htmlClient.onReady();
    context.subscriptions.push(activateAutoInsertion((document, position) => {
        const params = htmlClient.code2ProtocolConverter.asTextDocumentPositionParams(document, position);
        return htmlClient.sendRequest(shared.GetAutoQuoteEditsRequest.type, params);
    }, { vue: true }, 'html.autoCreateQuotes', (lastChange, lastCharacter) => lastChange.rangeLength === 0 && lastCharacter === '='));
    context.subscriptions.push(activateAutoInsertion((document, position) => {
        const params = htmlClient.code2ProtocolConverter.asTextDocumentPositionParams(document, position);
        return htmlClient.sendRequest(shared.GetTagCloseEditsRequest.type, params);
    }, { vue: true }, 'html.autoClosingTags', (lastChange, lastCharacter) => lastChange.rangeLength === 0 && (lastCharacter === '>' || lastCharacter === '/')));
    context.subscriptions.push(activateAutoInsertion((document, position) => {
        const params = tsClient.code2ProtocolConverter.asTextDocumentPositionParams(document, position);
        return tsClient.sendRequest(shared.GetRefCompleteEditsRequest.type, params);
    }, {
        vue: true,
        javascript: true,
        typescript: true,
        javascriptreact: true,
        typescriptreact: true,
    }, 'volar.autoCompleteRefs', (lastChange, lastCharacter, document) => {
        const rangeStart = lastChange.range.start;
        const position = new vscode.Position(rangeStart.line, rangeStart.character + lastChange.text.length);
        const nextCharacter = document.getText(new vscode.Range(position, document.positionAt(document.offsetAt(position) + 1)));
        if (lastCharacter === undefined) { // delete text
            return false;
        }
        if (lastChange.text.indexOf('\n') >= 0) { // multi-line change
            return false;
        }
        return /\w/.test(lastCharacter) && !/\w/.test(nextCharacter);
    }));
}
exports.activate = activate;
function activateAutoInsertion(provider, supportedLanguages, configName, changeValid) {
    let disposables = [];
    vscode.workspace.onDidChangeTextDocument(onDidChangeTextDocument, null, disposables);
    let isEnabled = false;
    updateEnabledState();
    vscode.window.onDidChangeActiveTextEditor(updateEnabledState, null, disposables);
    let timeout;
    function updateEnabledState() {
        isEnabled = false;
        let editor = vscode.window.activeTextEditor;
        if (!editor) {
            return;
        }
        let document = editor.document;
        if (!supportedLanguages[document.languageId]) {
            return;
        }
        if (!vscode.workspace.getConfiguration(undefined, document.uri).get(configName, true)) {
            return;
        }
        isEnabled = true;
    }
    function onDidChangeTextDocument({ document, contentChanges, reason }) {
        var _a;
        if (!isEnabled || contentChanges.length === 0 || reason === vscode.TextDocumentChangeReason.Undo || reason === vscode.TextDocumentChangeReason.Redo) {
            return;
        }
        const activeDocument = (_a = vscode.window.activeTextEditor) === null || _a === void 0 ? void 0 : _a.document;
        if (document !== activeDocument) {
            return;
        }
        if (timeout) {
            clearTimeout(timeout);
            timeout = undefined;
        }
        const lastChange = contentChanges[contentChanges.length - 1];
        const lastCharacter = lastChange.text[lastChange.text.length - 1];
        if (changeValid(lastChange, lastCharacter, document)) {
            doAutoInsert(document, lastChange);
        }
    }
    function doAutoInsert(document, lastChange) {
        const rangeStart = lastChange.range.start;
        const version = document.version;
        timeout = setTimeout(() => {
            const position = new vscode.Position(rangeStart.line, rangeStart.character + lastChange.text.length);
            provider(document, position).then(text => {
                if (text && isEnabled) {
                    const activeEditor = vscode.window.activeTextEditor;
                    if (activeEditor) {
                        const activeDocument = activeEditor.document;
                        if (document === activeDocument && activeDocument.version === version) {
                            const selections = activeEditor.selections;
                            if (selections.length && selections.some(s => s.active.isEqual(position))) {
                                activeEditor.insertSnippet(new vscode.SnippetString(text), selections.map(s => s.active));
                            }
                            else {
                                activeEditor.insertSnippet(new vscode.SnippetString(text), position);
                            }
                        }
                    }
                }
            });
            timeout = undefined;
        }, 100);
    }
    return vscode.Disposable.from(...disposables);
}
