// @volar/server/out/project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getScriptText = exports.createProject = exports.renameFileContentCache = exports.fileRenamings = void 0;
const shared = require("@volar/shared");
const vue = require("vscode-vue-languageservice");
const vscode = require("vscode-languageserver");
const schemaRequestService_1 = require("./schemaRequestService");
exports.fileRenamings = new Set();
exports.renameFileContentCache = new Map();
function createProject(ts, options, rootPath, tsConfig, tsLocalized, documents, workDoneProgress, connection, lsConfigs) {
    let typeRootVersion = 0;
    let tsProjectVersion = 0;
    let vueProjectVersion = 0;
    let parsedCommandLine;
    let vueLs;
    const scripts = new shared.FsPathMap();
    const languageServiceHost = createLanguageServiceHost();
    const disposables = [];
    init();
    return {
        onWorkspaceFilesChanged,
        onDocumentUpdated,
        getLanguageService,
        getLanguageServiceDontCreate: () => vueLs,
        getParsedCommandLine: () => parsedCommandLine,
        dispose,
    };
    function getLanguageService() {
        if (!vueLs) {
            let numOfFeatures = 0;
            if (options.languageFeatures) {
                for (let feature in options.languageFeatures) {
                    if (!!options.languageFeatures[feature]) {
                        numOfFeatures++;
                    }
                }
            }
            vueLs = vue.createLanguageService({ typescript: ts }, languageServiceHost);
            vueLs.__internal__.onInitProgress(p => {
                if (p === 0) {
                    workDoneProgress.begin(`Initializing Vue language features (${numOfFeatures} features)`);
                }
                if (p < 1) {
                    workDoneProgress.report(p * 100);
                }
                else {
                    workDoneProgress.done();
                }
            });
        }
        return vueLs;
    }
    async function onWorkspaceFilesChanged(changes) {
        await Promise.all([...exports.fileRenamings]);
        for (const change of changes) {
            const fileName = shared.uriToFsPath(change.uri);
            const script = scripts.get(fileName);
            if (script && change.type === vscode.FileChangeType.Changed) {
                if (script.version >= 0) {
                    script.version = -1;
                }
                else {
                    script.version--;
                }
            }
            else if (script && change.type === vscode.FileChangeType.Deleted) {
                scripts.delete(fileName);
            }
            updateProjectVersion(fileName);
        }
        const creates = changes.filter(change => change.type === vscode.FileChangeType.Created);
        const deletes = changes.filter(change => change.type === vscode.FileChangeType.Deleted);
        if (creates.length || deletes.length) {
            parsedCommandLine = createParsedCommandLine();
            typeRootVersion++; // TODO: check changed in node_modules?
        }
    }
    async function init() {
        await Promise.all([...exports.fileRenamings]);
        parsedCommandLine = createParsedCommandLine();
    }
    async function onDocumentUpdated(document) {
        await Promise.all([...exports.fileRenamings]);
        const fileName = shared.uriToFsPath(document.uri);
        const script = scripts.get(fileName);
        if (script) {
            script.version = document.version;
        }
        updateProjectVersion(fileName);
    }
    function updateProjectVersion(changedFileName) {
        if (changedFileName.endsWith('.vue')) {
            vueProjectVersion++;
        }
        else {
            tsProjectVersion++;
        }
    }
    function createLanguageServiceHost() {
        var _a;
        const host = {
            // vue
            createTsLanguageService(host) {
                return shared.createTsLanguageService(ts, host);
            },
            getEmmetConfig: lsConfigs.getEmmetConfiguration,
            schemaRequestService: ((_a = options.languageFeatures) === null || _a === void 0 ? void 0 : _a.schemaRequestService) ? (0, schemaRequestService_1.getSchemaRequestService)(connection, options.languageFeatures.schemaRequestService) : undefined,
            getPreferences: lsConfigs.getTsPreferences,
            getFormatOptions: lsConfigs.getTsFormatOptions,
            getParsedCommandLine: () => parsedCommandLine,
            getCssLanguageSettings: lsConfigs.getCssLanguageSettings,
            // ts
            getNewLine: () => ts.sys.newLine,
            useCaseSensitiveFileNames: () => ts.sys.useCaseSensitiveFileNames,
            readFile: ts.sys.readFile,
            writeFile: ts.sys.writeFile,
            directoryExists: ts.sys.directoryExists,
            getDirectories: ts.sys.getDirectories,
            readDirectory: ts.sys.readDirectory,
            realpath: ts.sys.realpath,
            fileExists: ts.sys.fileExists,
            // custom
            getDefaultLibFileName: options => ts.getDefaultLibFilePath(options),
            getProjectVersion: () => tsProjectVersion.toString(),
            getVueProjectVersion: () => vueProjectVersion.toString(),
            getTypeRootsVersion: () => typeRootVersion,
            getScriptFileNames: () => parsedCommandLine.fileNames,
            getCurrentDirectory: () => rootPath,
            getCompilationSettings: () => parsedCommandLine.options,
            getScriptVersion,
            getScriptSnapshot,
        };
        if (tsLocalized) {
            host.getLocalizedDiagnosticMessages = () => tsLocalized;
        }
        return host;
        function getScriptVersion(fileName) {
            var _a, _b;
            return (_b = (_a = scripts.get(fileName)) === null || _a === void 0 ? void 0 : _a.version.toString()) !== null && _b !== void 0 ? _b : '';
        }
        function getScriptSnapshot(fileName) {
            const script = scripts.get(fileName);
            if (script && script.snapshotVersion === script.version) {
                return script.snapshot;
            }
            const text = getScriptText(ts, documents, fileName);
            if (text !== undefined) {
                const snapshot = ts.ScriptSnapshot.fromString(text);
                if (script) {
                    script.snapshot = snapshot;
                    script.snapshotVersion = script.version;
                }
                else {
                    scripts.set(fileName, {
                        version: -1,
                        snapshot: snapshot,
                        snapshotVersion: -1,
                    });
                }
                return snapshot;
            }
        }
    }
    function dispose() {
        if (vueLs) {
            vueLs.dispose();
        }
        for (const disposable of disposables) {
            disposable.dispose();
        }
        scripts.clear();
        disposables.length = 0;
    }
    function createParsedCommandLine() {
        const parseConfigHost = {
            ...ts.sys,
            readDirectory: (path, extensions, exclude, include, depth) => {
                return ts.sys.readDirectory(path, [...extensions, '.vue'], exclude, include, depth);
            },
        };
        if (typeof tsConfig === 'string') {
            return shared.createParsedCommandLine(ts, parseConfigHost, tsConfig);
        }
        else {
            const content = ts.parseJsonConfigFileContent({}, parseConfigHost, rootPath, tsConfig, 'tsconfig.json');
            content.options.outDir = undefined; // TODO: patching ts server broke with outDir + rootDir + composite/incremental
            content.fileNames = content.fileNames.map(shared.normalizeFileName);
            return content;
        }
    }
}
exports.createProject = createProject;
function getScriptText(ts, documents, fileName) {
    const doc = documents.get(shared.fsPathToUri(fileName));
    if (doc) {
        return doc.getText();
    }
    if (ts.sys.fileExists(fileName)) {
        return ts.sys.readFile(fileName, 'utf8');
    }
    return exports.renameFileContentCache.get(shared.fsPathToUri(fileName));
}
exports.getScriptText = getScriptText;
