// @volar/server/out/features/languageFeatures.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const shared = require("@volar/shared");
const vue = require("vscode-vue-languageservice");
const vscode = require("vscode-languageserver");
const project_1 = require("../project");
function register(ts, connection, documents, getProjects, features, lsConfigs, params) {
    connection.onCompletion(async (handler) => {
        var _a, _b, _c, _d, _e, _f;
        const list = await ((_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(handler.textDocument.uri)) === null || _b === void 0 ? void 0 : _b.service.doComplete(handler.textDocument.uri, handler.position, handler.context, () => connection.workspace.getConfiguration('volar.completion.autoImportComponent'), async (uri) => {
            var _a;
            if ((_a = features.completion) === null || _a === void 0 ? void 0 : _a.getDocumentNameCasesRequest) {
                return await connection.sendRequest(shared.GetDocumentNameCasesRequest.type, { uri });
            }
            return {
                tagNameCase: features.completion.defaultTagNameCase,
                attrNameCase: features.completion.defaultAttrNameCase,
            };
        }));
        const insertReplaceSupport = (_f = (_e = (_d = (_c = params.capabilities.textDocument) === null || _c === void 0 ? void 0 : _c.completion) === null || _d === void 0 ? void 0 : _d.completionItem) === null || _e === void 0 ? void 0 : _e.insertReplaceSupport) !== null && _f !== void 0 ? _f : false;
        if (!insertReplaceSupport && list) {
            for (const item of list.items) {
                if (item.textEdit && vscode.InsertReplaceEdit.is(item.textEdit)) {
                    item.textEdit = vscode.TextEdit.replace(item.textEdit.insert, item.textEdit.newText);
                }
            }
        }
        return list;
    });
    connection.onCompletionResolve(async (item) => {
        var _a, _b, _c, _d, _e;
        const uri = (_a = item.data) === null || _a === void 0 ? void 0 : _a.uri;
        if (!uri)
            return item;
        const activeSel = ((_b = features.completion) === null || _b === void 0 ? void 0 : _b.getDocumentSelectionRequest)
            ? await connection.sendRequest(shared.GetEditorSelectionRequest.type)
            : undefined;
        const newPosition = (activeSel === null || activeSel === void 0 ? void 0 : activeSel.textDocument.uri.toLowerCase()) === uri.toLowerCase() ? activeSel.position : undefined;
        return (_e = (_d = (_c = getProjects()) === null || _c === void 0 ? void 0 : _c.get(uri)) === null || _d === void 0 ? void 0 : _d.service.doCompletionResolve(item, newPosition)) !== null && _e !== void 0 ? _e : item;
    });
    connection.onHover(async (handler) => {
        var _a, _b;
        return await ((_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(handler.textDocument.uri)) === null || _b === void 0 ? void 0 : _b.service.doHover(handler.textDocument.uri, handler.position));
    });
    connection.onSignatureHelp(handler => {
        var _a, _b;
        return (_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(handler.textDocument.uri)) === null || _b === void 0 ? void 0 : _b.service.getSignatureHelp(handler.textDocument.uri, handler.position, handler.context);
    });
    connection.onPrepareRename(handler => {
        var _a, _b;
        return (_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(handler.textDocument.uri)) === null || _b === void 0 ? void 0 : _b.service.prepareRename(handler.textDocument.uri, handler.position);
    });
    connection.onRenameRequest(async (handler) => {
        var _a, _b;
        return await ((_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(handler.textDocument.uri)) === null || _b === void 0 ? void 0 : _b.service.doRename(handler.textDocument.uri, handler.position, handler.newName));
    });
    connection.onCodeLens(async (handler) => {
        var _a, _b;
        return (_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(handler.textDocument.uri)) === null || _b === void 0 ? void 0 : _b.service.getCodeLens(handler.textDocument.uri, await (lsConfigs === null || lsConfigs === void 0 ? void 0 : lsConfigs.getCodeLensConfigs()));
    });
    connection.onCodeLensResolve(codeLens => {
        var _a, _b, _c, _d;
        const uri = (_a = codeLens.data) === null || _a === void 0 ? void 0 : _a.uri;
        return (_d = (_c = (_b = getProjects()) === null || _b === void 0 ? void 0 : _b.get(uri)) === null || _c === void 0 ? void 0 : _c.service.doCodeLensResolve(codeLens, typeof features.codeLens === 'object' && features.codeLens.showReferencesNotification)) !== null && _d !== void 0 ? _d : codeLens;
    });
    connection.onExecuteCommand(handler => {
        var _a, _b, _c;
        const uri = (_a = handler.arguments) === null || _a === void 0 ? void 0 : _a[0];
        return (_c = (_b = getProjects()) === null || _b === void 0 ? void 0 : _b.get(uri)) === null || _c === void 0 ? void 0 : _c.service.__internal__.executeCommand(uri, handler.command, handler.arguments, connection);
    });
    connection.onCodeAction(async (handler) => {
        var _a;
        const uri = handler.textDocument.uri;
        const project = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(uri);
        if (project) {
            const codeActions = await project.service.getCodeActions(uri, handler.range, handler.context);
            for (const codeAction of codeActions) {
                if (codeAction.data && typeof codeAction.data === 'object') {
                    codeAction.data.uri = uri;
                }
                else {
                    codeAction.data = { uri };
                }
            }
            return codeActions;
        }
    });
    connection.onCodeActionResolve(async (codeAction) => {
        var _a, _b;
        const uri = (_a = codeAction.data) === null || _a === void 0 ? void 0 : _a.uri;
        const project = uri ? (_b = getProjects()) === null || _b === void 0 ? void 0 : _b.get(uri) : undefined;
        if (project) {
            return await project.service.doCodeActionResolve(codeAction);
        }
        return codeAction;
    });
    connection.onReferences(async (handler) => {
        var _a, _b;
        return await ((_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(handler.textDocument.uri)) === null || _b === void 0 ? void 0 : _b.service.findReferences(handler.textDocument.uri, handler.position));
    });
    connection.onDefinition(async (handler) => {
        var _a, _b;
        return await ((_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(handler.textDocument.uri)) === null || _b === void 0 ? void 0 : _b.service.findDefinition(handler.textDocument.uri, handler.position));
    });
    connection.onTypeDefinition(handler => {
        var _a, _b;
        return (_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(handler.textDocument.uri)) === null || _b === void 0 ? void 0 : _b.service.findTypeDefinition(handler.textDocument.uri, handler.position);
    });
    connection.onDocumentHighlight(handler => {
        var _a, _b;
        return (_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(handler.textDocument.uri)) === null || _b === void 0 ? void 0 : _b.service.findDocumentHighlights(handler.textDocument.uri, handler.position);
    });
    connection.onDocumentLinks(async (handler) => {
        var _a, _b;
        return await ((_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(handler.textDocument.uri)) === null || _b === void 0 ? void 0 : _b.service.findDocumentLinks(handler.textDocument.uri));
    });
    connection.onWorkspaceSymbol(async (handler, token) => {
        const projects = getProjects();
        if (projects) {
            const _projects = projects.projects.size ? projects.projects : projects.inferredProjects;
            let results = [];
            for (const [_, project] of _projects) {
                if (token.isCancellationRequested)
                    return;
                results = results.concat(await project.getLanguageService().findWorkspaceSymbols(handler.query));
            }
            return results;
        }
    });
    connection.languages.callHierarchy.onPrepare(async (handler) => {
        var _a, _b;
        const items = await ((_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(handler.textDocument.uri)) === null || _b === void 0 ? void 0 : _b.service.callHierarchy.doPrepare(handler.textDocument.uri, handler.position));
        if (items) {
            for (const item of items) {
                if (typeof item.data !== 'object')
                    item.data = {};
                item.data.__uri = handler.textDocument.uri;
            }
        }
        return (items === null || items === void 0 ? void 0 : items.length) ? items : null;
    });
    connection.languages.callHierarchy.onIncomingCalls(handler => {
        var _a, _b, _c, _d;
        const data = handler.item.data;
        const uri = (_a = data === null || data === void 0 ? void 0 : data.__uri) !== null && _a !== void 0 ? _a : handler.item.uri;
        return (_d = (_c = (_b = getProjects()) === null || _b === void 0 ? void 0 : _b.get(uri)) === null || _c === void 0 ? void 0 : _c.service.callHierarchy.getIncomingCalls(handler.item)) !== null && _d !== void 0 ? _d : [];
    });
    connection.languages.callHierarchy.onOutgoingCalls(handler => {
        var _a, _b, _c, _d;
        const data = handler.item.data;
        const uri = (_a = data === null || data === void 0 ? void 0 : data.__uri) !== null && _a !== void 0 ? _a : handler.item.uri;
        return (_d = (_c = (_b = getProjects()) === null || _b === void 0 ? void 0 : _b.get(uri)) === null || _c === void 0 ? void 0 : _c.service.callHierarchy.getOutgoingCalls(handler.item)) !== null && _d !== void 0 ? _d : [];
    });
    connection.languages.semanticTokens.on(async (handler, token, _, resultProgress) => {
        var _a, _b, _c;
        const result = await ((_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(handler.textDocument.uri)) === null || _b === void 0 ? void 0 : _b.service.getSemanticTokens(handler.textDocument.uri, undefined, token, resultProgress));
        return {
            resultId: result === null || result === void 0 ? void 0 : result.resultId,
            data: (_c = result === null || result === void 0 ? void 0 : result.data) !== null && _c !== void 0 ? _c : [],
        };
    });
    connection.languages.semanticTokens.onRange(async (handler, token, _, resultProgress) => {
        var _a, _b, _c;
        const result = await ((_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(handler.textDocument.uri)) === null || _b === void 0 ? void 0 : _b.service.getSemanticTokens(handler.textDocument.uri, handler.range, token, resultProgress));
        return {
            resultId: result === null || result === void 0 ? void 0 : result.resultId,
            data: (_c = result === null || result === void 0 ? void 0 : result.data) !== null && _c !== void 0 ? _c : [],
        };
    });
    connection.workspace.onWillRenameFiles(async (handler) => {
        const hasTsFile = handler.files.some(file => file.newUri.endsWith('.vue') || file.newUri.endsWith('.ts') || file.newUri.endsWith('.tsx'));
        const config = await connection.workspace.getConfiguration(hasTsFile ? 'typescript.updateImportsOnFileMove.enabled' : 'javascript.updateImportsOnFileMove.enabled');
        if (config === 'always') {
            const renaming = new Promise(async (resolve) => {
                var _a, _b;
                for (const file of handler.files) {
                    const renameFileContent = (0, project_1.getScriptText)(ts, documents, shared.uriToFsPath(file.oldUri));
                    if (renameFileContent) {
                        project_1.renameFileContentCache.set(file.oldUri, renameFileContent);
                    }
                }
                await shared.sleep(0);
                const edit = await worker();
                if (edit) {
                    if (edit.documentChanges) {
                        for (const change of edit.documentChanges) {
                            if (vscode.TextDocumentEdit.is(change)) {
                                for (const file of handler.files) {
                                    if (change.textDocument.uri === file.oldUri) {
                                        change.textDocument.uri = file.newUri;
                                        change.textDocument.version = (_b = (_a = documents.get(file.newUri)) === null || _a === void 0 ? void 0 : _a.version) !== null && _b !== void 0 ? _b : change.textDocument.version;
                                    }
                                }
                            }
                        }
                    }
                    connection.workspace.applyEdit(edit);
                }
                resolve();
            });
            project_1.fileRenamings.add(renaming);
            (async () => {
                await renaming;
                project_1.fileRenamings.delete(renaming);
                project_1.renameFileContentCache.clear();
            })();
        }
        if (config === 'prompt')
            return await worker();
        return null;
        async function worker() {
            const edits = (await Promise.all(handler.files
                .map(async (file) => {
                var _a, _b;
                return await ((_b = (_a = getProjects()) === null || _a === void 0 ? void 0 : _a.get(file.oldUri)) === null || _b === void 0 ? void 0 : _b.service.getEditsForFileRename(file.oldUri, file.newUri));
            }))).filter(shared.notEmpty);
            if (edits.length) {
                const result = edits[0];
                vue.margeWorkspaceEdits(result, ...edits.slice(1));
                return result;
            }
            return null;
        }
    });
}
exports.register = register;
